### PR TITLE
refactor(mqtt3): cleanup MQTTClient

### PR DESCRIPTION
**Pull request by jcosentino11**
**Issue #, if available:**

**Description of changes:**
Refactor methods, adjust naming. Add an integration test for MQTTClient 

**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java [hunk 1]
```diff
@@ -252,84 +259,115 @@ private MqttConnectOptions getConnectionOptions() throws KeyStoreException {
...
-
-        connectFuture = executorService.submit(this::reconnectAndResubscribe);
+    private void connectAndSubscribeAsync() {
+        synchronized (connectLock) {
+            if (connectFuture != null) {
+                connectFuture.cancel(true);
+            }
+            connectFuture = executorService.submit(() -> {
+                try {
+                    connect();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:297]: is this required? Doesn't the client automatically reconnect when needed?

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java [hunk 2]
```diff
@@ -252,84 +259,115 @@ private MqttConnectOptions getConnectionOptions() throws KeyStoreException {
...
     }
 
-    private synchronized void doConnect() throws MqttException, KeyStoreException {
-        if (!mqttClientInternal.isConnected()) {
-            mqttClientInternal.connect(getConnectionOptions());
+    @SuppressWarnings({
+            "PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException", "PMD.ExceptionAsFlowControl"})
+    private void connect() throws InterruptedException {
+        if (mqttClientInternal.isConnected()) {
             LOGGER.atInfo()
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:310]: too much is being logged at info

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java (hunk 1 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:297]: there's an option in paho for auto reconnect but we don't use that afaik

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java [hunk 3]
```diff
@@ -252,84 +259,115 @@ private MqttConnectOptions getConnectionOptions() throws KeyStoreException {
...
+                return null;
+            }, "connect-mqtt-topic", LOGGER);
+            LOGGER.atInfo()
+                    .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
+                    .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
+                    .log("Connected to broker");
+        } catch (InterruptedException e) {
+            throw e;
+        } catch (Exception e) {
+            // TODO considering erroring the component, or other recovery options
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:344]: retrying?

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:310]: will audit the class and adjust levels

**Review by MikeDombo — COMMENTED** [review]
Looks like we aren't using persistent sessions, is that right? We should be able to save a lot of time when reconnecting if the session was persisted

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java (hunk 1 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:297]: Why not? Should we?

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java (hunk 3 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:344]: yeah that's fair, i suppose doesn't hurt to retry everything

**Comment by jcosentino11** [comment]
> Looks like we aren't using persistent sessions, is that right? We should be able to save a lot of time when reconnecting if the session was persisted

yeah, was just going on original impl that didn't use persistent sessions so that wasn't in scope for me.   i think we may also need to switch from inmemorypersistence if we do persisten sessions? could do this as a follow up

**Comment by MikeDombo** [comment]
> > Looks like we aren't using persistent sessions, is that right? We should be able to save a lot of time when reconnecting if the session was persisted
> 
> yeah, was just going on original impl that didn't use persistent sessions so that wasn't in scope for me. i think we may also need to switch from inmemorypersistence if we do persistent sessions? could do this as a follow up

No, the session I'm talking about is in the broker itself, not the client.

**Comment by jcosentino11** [comment]
> > > Looks like we aren't using persistent sessions, is that right? We should be able to save a lot of time when reconnecting if the session was persisted
> > 
> > 
> > yeah, was just going on original impl that didn't use persistent sessions so that wasn't in scope for me. i think we may also need to switch from inmemorypersistence if we do persistent sessions? could do this as a follow up
> 
> No, the session I'm talking about is in the broker itself, not the client.

yeah-- ok i'll give this another pass with the stuff you suggested, should simplify things alot

**Comment by jcosentino11** [comment]
Since all the async changes were included in 2.3.0 release, turned this into a refactor PR.  Will explore using cleanSession=false in a separate PR, although it is nice to not have to unsubscribe on shutdown :)

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java [hunk 4]
```diff
@@ -228,16 +239,19 @@ public void publish(MqttMessage message) throws MessageClientException {
...
         if (mqttClientInternal.isConnected()) {
-            updateSubscriptionsInternal();
+            updateSubscriptionsAsync();
         }
     }
 
     @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    private void updateSubscriptionsInternal() {
+    private void updateSubscriptionsAsync() {
         synchronized (subscribeLock) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:255]: this doesn't look async. We're unsubscribing from all topics synchronously. This method also then doesn't subscribe to any topics?

File: src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java [hunk 5]
```diff
@@ -42,11 +42,27 @@
...
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(MqttException.class))
+                    .build();
+    private static final RetryUtils.RetryConfig RETRY_CONFIG_CONNECT =
+            RetryUtils.RetryConfig.builder()
+                    .initialRetryInterval(Duration.ofSeconds(1L))
+                    .maxRetryInterval(Duration.ofSeconds(120L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    // always attempt to retry
+                    .retryableExceptions(Collections.singletonList(Exception.class))
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqtt/bridge/clients/MQTTClient.java:58]: are we going to end up retrying on interrupted?

**Pull request closed without merging**